RFCs/0035-partiql-window-function.md

```diff
@@ -0,0 +1,1204 @@
+
+This RFC purposes the semantics for **window partition** in PartiQL. Subsequent RFCs are expected to cover other concepts in window function such as **peer** and **window frame**.
+
+## Guide-level explanation
+
+### Out of scope
+
+* This RFC focus on explaining the semantics of PartiQL’s window functions. The PartiQL core syntax provided in this RFC is for guidance only. Finalizing grammar and syntax is out of the scope, excluding the supported SQL compatible window function. 
+* SQL’s window implicitly defines three concepts, partition, frame, and peer. As a first step, this RFC focus on partition only, the concept of frame and peer are out of scope for now, but some preliminary work has been done to make sure the other concepts can fit into the framework in a modular fashion.
+* An exhaustive mapping between SQL’s window function and PartiQL’s expression is out of the scope. In this RFC, we only define the mapping between SQL’s `lag` function and it's corresponding PartiQL expression as an example. The mapping between other functions are skipped.
+
+We assume that the readers of this RFC has basic understanding of SQL’s window function and PartiQL semantics.
+
+### Terminology
+
+1. RFC key word: MUST, MUST NOT, REQUIRED, SHALL, SHALL NOT, etc. See Appendix (1) for detail. 
+2. Windowed Table and Window: Windowed table is a defined in SQL spec 2011, section 4.15.14. A **windowed table** is a table together with one or more windows. A **window** is a transient data structure associated
+   with a table expression. A window is defined explicitly by a window definition or implicitly by an inline window specification. Implicitly defined windows have an implementation-dependent window name. A
+   **window** is used to specify window partitions and window frames, which are collections of rows used in the definition of **window functions**.
+3. Partition: To distinguish between SQL's definition of window, in this RFC, the term **partition** refers to the window partition specified for each row. See [SQL’s Windowed Table As Nested Data](#SQL’s-Windowed-Table-As-Nested-Data) for details. 
+4. Equivalent Partition: Two binding tuples $b, b' \in B^{in}_{WINDOWED}$ are in the same equivalence partition if and only if every partition expression $e_i$ evaluates to equivalent values $v_i$ (when evaluted on b) and $v_i'$ (when evaluated on $b'$). 
+
+### SQL’s Windowed Table As Nested Data
+In this section, we introduce a way to model SQL's windowed table as nested data.
+Notice that the nested data format, although not fitting in SQL's formal semantics, provides a simple and intuitive way to understand window function results, and forms the basis for PartiQL's window semantics. 
+
+SQL’s in-line window function that operates without frame clause looks like:
+
+```sql
```

Comment:
Can we also exemplify this (E.g. adding the syntax for `LAG`)

Reply:
Do you mean: 

```
For example, the syntax for `LAG` function in SQL is: 

LAG (expression [,offset] [,default])  
   OVER ( 
            [PARTITION BY <partition_expression>] 
            ORDER BY <sort spec> )
```

I was a little concerned about adding this will raised the question on "what is lag function" or "what is offset" or "why is ORDER BY not optional for `LAG` here", those questions are answered in later section [Using SQL’s inline window function in PartiQL].

Reply:
Yes, and we can link to external documentation or mention that there will be more context in the upcoming sections.

Reply:
Addressed in the new commit.